### ISSUE TITLE
Using the 1.1 DCAP client without having $HOME defined causes a deadlock

### ISSUE THREAD

**Issue by achamayou** [microsoft/Azure-DCAP-Client#84]
Trace below. This isn't right, at worst an error should be returned, and at best it should proceed without the cache.

```
#0  0x00007f65e325d7e7 in __pthread_once_slow () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f65e183ef39 in ?? () from /usr/lib/libdcap_quoteprov.so
#2  0x00007f65e1840935 in void std::call_once<void (&)()>(std::once_flag&, void (&)()) () from /usr/lib/libdcap_quoteprov.so
#3  0x00007f65e183f698 in ?? () from /usr/lib/libdcap_quoteprov.so
#4  0x00007f65e183fdf1 in local_cache_get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from /usr/lib/libdcap_quoteprov.so
#5  0x00007f65e1833dd6 in sgx_ql_get_quote_config () from /usr/lib/libdcap_quoteprov.so
#6  0x00007f65e3d63889 in ?? () from /usr/lib/x86_64-linux-gnu/libsgx_dcap_ql.so.1
#7  0x00007f65e3d6495b in ?? () from /usr/lib/x86_64-linux-gnu/libsgx_dcap_ql.so.1
#8  0x00007f65e3d6337d in ?? () from /usr/lib/x86_64-linux-gnu/libsgx_dcap_ql.so.1
#9  0x0000000000475492 in oe_sgx_qe_get_quote_size (quote_size=0x7f65e1a4a620) at ../host/sgx/sgxquote.c:32
#10 0x0000000000471f7a in sgx_get_quote_size (quote_size=<optimized out>) at ../host/sgx/quote.c:178
#11 sgx_get_quote (report=<optimized out>, quote=<optimized out>, quote_size=<optimized out>) at ../host/sgx/quote.c:201
#12 0x00000000004718f9 in HandleGetQuote (arg_in=<optimized out>) at ../host/sgx/ocalls.c:135
#13 0x000000000046ad50 in _handle_ocall (enclave=<optimized out>, tcs=<optimized out>, func=<optimized out>, arg_in=<optimized out>, arg_out=<optimized out>) at ../host/sgx/calls.c:485
#14 __oe_dispatch_ocall (arg1=985171008421888, arg2=<optimized out>, arg1_out=<optimized out>, arg2_out=<optimized out>, tcs_=<optimized out>, enclave=<optimized out>) at ../host/sgx/calls.c:606
#15 0x000000000047b058 in __oe_host_stack_bridge () at ../host/sgx/linux/enter.S:271
#16 0x000000000047afa0 in __morestack () at ../host/sgx/linux/enter.S:144
#17 0x000000000046b8d6 in _do_eenter (enclave=<optimized out>, tcs=<optimized out>, aep=<optimized out>, code_in=OE_CODE_ECALL, arg_in=<optimized out>, func_in=<optimized out>,
    code_out=<optimized out>, func_out=<optimized out>, result_out=<optimized out>, arg_out=<optimized out>) at ../host/sgx/calls.c:228
#18 oe_ecall (enclave=<optimized out>, func=<optimized out>, arg=<optimized out>, arg_out_ptr=<optimized out>) at ../host/sgx/calls.c:775
#19 0x000000000046bc36 in oe_call_enclave_function_by_table_id (enclave=0x7f65e1a4a620, table_id=<optimized out>, function_id=<optimized out>, input_buffer=<optimized out>, input_buffer_size=202,
    output_buffer=0x1, output_buffer_size=<optimized out>, output_bytes_written=<optimized out>) at ../host/sgx/calls.c:852
#20 0x000000000046be06 in oe_call_enclave_function (enclave=0x7f65e1a4a620, function_id=<optimized out>, input_buffer=0x7f65e325d7e7 <__pthread_once_slow+119>, input_buffer_size=202,
    output_buffer=0x1, output_buffer_size=0, output_bytes_written=<optimized out>) at ../host/sgx/calls.c:890
#21 0x000000000046a603 in enclave_create_node (enclave=0x89c640, _retval=0x7ffd8d646e07, config=<optimized out>, node_cert=0x10a0030 "", node_cert_size=4096, node_cert_len=0x7ffd8d646e18,
    quote=0x10a1040 "", quote_size=10240, quote_len=0x7ffd8d646e10, recover=<optimized out>) at ccf_u.cpp:91
#22 0x00000000004157dc in host::Enclave::create_node (this=<optimized out>, config=..., node_cert=std::vector of length 4096, capacity 4096 = {...},
    quote=std::vector of length 10240, capacity 10240 = {...}, recover=false) at ../src/host/enclave.h:88
#23 0x000000000040be69 in main (argc=9025584, argv=0x1000000) at ../src/host/main.cpp:252
```

**Comment by achamayou** [comment]
To clarify, 1.0 does not exhibit this behaviour.

**Comment by letmaik** [comment]
I'm hitting the same, in my case because running this inside a Docker container where $HOME=/ and then it fails with:
> dcap_quoteprov: [ERROR]: Unknown exception thrown, error: Error creating directory '//.az-dcap-client/': Permission denied

@msft-jonnadul

**Comment by wintersteiger** [comment]
+1

**Comment by msft-jonnadul** [comment]
Ah interesting, just for context the .azure-dcap-client/ folder is created to house the local cache of pck certificates for faster quote generation. This was not an issue in 1.0 because local caching was shipped with 1.1.

Currently the logic in the Azure DCAP is such that if a AZDCAP_CACHE environment variable is specified then it'll house the local cache there if not it'll default to HOME. If HOME is undefined, then it'll actually throw a runtime exception but in this case it is defined but to /, which invoked this permissions issue.

To unblock, you can pass in the AZDCAP_CACHE environment variable to some permissible location into your Docker instance. We can work on a fix where if HOME is undefined or set to / we'll just not attempt to do any local caching.

**Comment by deschuma** [comment]
This is addressed here: https://github.com/microsoft/Azure-DCAP-Client/pull/97

**Comment by deschuma** [comment]
@achamayou do you mind rechecking and closing with latest?

**Comment by achamayou** [comment]
@deschuma I can confirm that this is fixed in 1.4.
